lib/BloqadeSchema/src/validate.jl

```diff
@@ -1,24 +1,122 @@
 
 
+# helper functions to generate violations
+
+message(::typeof(>)) = "exceeds maximum"
+message(::typeof(<)) = "below minimum"
+message(::typeof(!=)) = "is not equal to the"
 
 # checks of x is integer multiple of res
 # if it is return false otherwise return true
-function check_resolution(res::Float64,x::Float64)
+function check_resolution(res::Real,x::Real) 
     x == 0 && return false
 
     x_div_res = (abs(x) / res)
     return !isapprox(round(x_div_res),x_div_res)
 end
 
+function group_by_key(itr,key)
+    groups = Dict{Any,Set}()
+
+    foreach(enumerate(itr)) do (i,value)
+        key_value = value[key]
+        if haskey(groups,key_value)
+            push!(groups[key_value],i)
+        else
+            groups[key_value] = Set([i])
+        end
+    end
+
+    return groups
+end
+
 # TODO: implement this test based on validation in TaskManager
-validate_lattice(positions,::DeviceCapabilities) = Set([])
+function validate_lattice(positions,dc::DeviceCapabilities)
+
+    positions = map(positions) do position
```

Comment:
Are we expecting position to be a tuple here? Maybe it'd clearer to have a type annotation.

Reply:
But isn't Vector is the native type for json?